File: waymark4/server/collab.py
"""waymark-relay/2 (design §5): per-field revisions on the wire.

v1's relay shipped whole-value frames and left merge discipline to client
debounce timers — which dropped remote edits the moment a field was focused
(commit 41b0eac). Relay/2 makes the merge discipline protocol:

  server → joiner   {"type": "state", "values", "revs", "saved_at",
                     "stale", "participants"}
  client → server   {"type": "update", "field", "value", "base_rev"}
                    (``null`` value clears the field; ``base_rev`` is the
                    rev this edit was based on)
  server → sender   {"type": "saved", "field", "rev"}
  server → others   {"type": "update", "field", "value", "rev", "actor",
                     "saved_at"}
  server → sender   {"type": "reject", "field", "rev", "value"}
                    (base_rev was stale: here is the field's truth — apply
                    it, then re-edit)
  server → room     {"type": "presence", "event": "joined"|"left",
                     "actor", "participants"}
  server → room     {"type": "closed", "reason": "consumed"|"discarded"
                     |"gone"} — then the socket closes
  server → sender   {"type": "error", "errors": {field: [msgs]}}

Merge discipline: server-ordered per-field last-write-wins with explicit
staleness rejection — modest, declared, and enforced here rather than
emergent from two debounce timers. Character-level merge (CRDT/OT) is a
different ``protocol`` token; the drain rule and the revs map bind those
too.

The drain rule is inherited, not re-implemented: every accepted update goes
through :meth:`DraftStore.save_fields` — the same code as a plain draft
PUT. Rooms are local; liveness crosses workers via the bus (design §7):
accepted frames publish on ``waymark4_collab``, every worker's rooms
deliver to their local members, and the publisher skips its own echo.

Joining requires the parent action to render in ``actions`` for the
principal (design §11) — v1 gated on resource visibility only.
"""
from __future__ import annotations

import asyncio
import contextlib
import json
import logging
from typing import Any

from .drafts import COLLAB_CHANNEL, PROTOCOL, Audience, audience_of

log = logging.getLogger("waymark4.collab")

__all__ = ["PROTOCOL", "Audience", "CollabRooms", "serve", "actor_of"]


def actor_of(principal: Any) -> dict[str, str]:
    return {"id": principal.id, "type": principal.type,
            "display": principal.display or principal.id}


RoomKey = tuple[str, str, str, str]  # (kind, resource_id, action, part_key)


class Room:
    def __init__(self) -> None:
        self.lock = asyncio.Lock()  # serializes drains for one draft
        self.members: dict[Any, Any] = {}  # websocket → Principal


class CollabRooms:
    """Local rooms + the bus. ``broadcast`` is local-only; ``announce`` (via
    DraftStore) is what crosses workers."""

    def __init__(self, engine: Any) -> None:
        self.engine = engine
        self.rooms: dict[RoomKey, Room] = {}
        self._regate_sub: Any = None
        self._regate_task: Any = None
        engine.bus.listen(COLLAB_CHANNEL, self._on_bus)

    # ── affordance regate (design §2) ───────────────────────────────────
    # The join gate is re-run on affordance-changing transitions: a grant
    # revocation or a state change drops lapsed members' sockets *now*, not
    # when the room happens to close. v2's caveat ("a principal whose
    # affordance lapses keeps the socket") is structurally gone.
    def start_regate(self, dispatcher: Any) -> None:
        self._regate_sub = dispatcher.subscribe(
            classes=frozenset({"transition"}))
        self._regate_task = asyncio.create_task(
            self._regate_loop(), name="waymark4-collab-regate")

    async def stop_regate(self) -> None:
        if self._regate_task is not None:
            self._regate_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._regate_task
            self._regate_task = None

    async def _regate_loop(self) -> None:
        while True:
            t = await self._regate_sub.queue.get()
            try:
                if t.kind == "grant":
                    # a grant change can affect any room's token principals
                    for key in list(self.rooms):
                        await self._regate_room(key)
                else:
                    for key in [k for k in self.rooms
                                if k[0] == t.kind and k[1] == t.resource_id]:
                        await self._regate_room(key)
            except Exception:  # a regate failure must not kill the loop
                log.exception("collab regate failed for transition %s", t.id)

    async def _regate_room(self, key: RoomKey) -> None:
        from .render import probe_transition

        room = self.rooms.get(key)
        if room is None or not room.members:
            return
        rdef = self.engine.registry.get(key[0])
        defn = rdef.machine.actions.get(key[2]) if rdef else None
        if rdef is None or defn is None:
            return
        async with self.engine.storage.session() as s:
            instance = await self.engine.storage.load(s, key[0], key[1])
            if instance is None:
                await self._close_local(key, "gone")
                return
            for ws, principal in list(room.members.items()):
                lapsed = instance.state not in defn.from_
                if not lapsed and getattr(principal, "scope", None) is not None:
                    # re-resolve the grant: the handshake's copy may be stale
                    from .grants import action_mode

                    grant = await self.engine.storage.load(
                        s, "grant", principal.scope.id)
                    lapsed = (grant is None or action_mode(
                        grant, self.engine.invoker.clock(), rdef.kind,
                        defn.name) == "none")
                if not lapsed:
                    ctx = self.engine.invoker._ctx(principal, s, mode="probe")
                    status, _, _, _ = await probe_transition(defn, instance, ctx)
                    lapsed = status != "available"
                if lapsed:
                    room.members.pop(ws, None)
                    with contextlib.suppress(Exception):
                        await ws.send_text(json.dumps(
                            {"type": "closed", "reason": "affordance lapsed"}))
                        await ws.close(code=4403)
        if room.members:
            await self.broadcast(key, {
                "type": "presence", "event": "left",
                "participants": self.participants(key)})

    def room(self, key: RoomKey) -> Room:
        return self.rooms.setdefault(key, Room())

    def participants(self, key: RoomKey) -> list[dict[str, str]]:
        room = self.rooms.get(key)
        return [actor_of(p) for p in room.members.values()] if room else []

    async def _on_bus(self, message: dict[str, Any]) -> None:
        """A frame from another worker (or our own echo, skipped)."""
        if message.get("_origin") == self.engine.bus.origin:
            return
        frame = message.get("frame")
        if not isinstance(frame, dict):
            return
        key = (message.get("kind"), message.get("id"),
               message.get("action"), message.get("part") or "")
        if frame.get("type") == "closed":
            await self._close_local(key, frame.get("reason", "gone"))
        else:
            await self.broadcast(key, frame)

    async def broadcast(self, key: RoomKey, message: dict[str, Any], *,
                        exclude: Any = None) -> None:
        room = self.rooms.get(key)
        if room is None:
            return
        payload = json.dumps(message, default=str)
        for ws in list(room.members):
            if ws is exclude:
                continue
            try:
                await ws.send_text(payload)
            except Exception:
                room.members.pop(ws, None)

    async def close(self, key: RoomKey, reason: str) -> None:
        """The draft is gone (consumed, discarded, resource vanished): tell
        every worker's room, then hang up locally."""
        rdef = self.engine.registry.get(key[0])
        defn = rdef.machine.actions.get(key[2]) if rdef else None
        if rdef is not None and defn is not None:
            await self.engine.draft_store.announce(
                rdef=rdef, defn=defn, resource_id=key[1], part_key=key[3],
                frame={"type": "closed", "reason": reason})
        await self._close_local(key, reason)

    async def _close_local(self, key: RoomKey, reason: str) -> None:
        room = self.rooms.pop(key, None)
        if room is None:
            return
        payload = json.dumps({"type": "closed", "reason": reason})
        for ws in list(room.members):
            try:
                await ws.send_text(payload)
                await ws.close()
            except Exception:
                pass


async def _drain(engine: Any, rdef: Any, defn: Any, resource_id: str,
                 part_key: str, field: str, value: Any, base_rev: int,
                 principal: Any, room: Room) -> dict[str, Any] | None:
    """One field update through the one write path. Returns
    {saved|reject...} or None when the resource is gone."""
    async with room.lock:
        async with engine.storage.session() as s:
            instance = await engine.storage.load(s, rdef.kind, resource_id)
            if instance is None:
                return None
            audience = audience_of(defn, principal)
            row = await engine.draft_store.load(
                s, rdef, defn, resource_id, part_key, audience)
            current_rev = int((row or {}).get("revs", {}).get(field, 0))
            if base_rev < current_rev:
                # stale edit: reject with the field's truth — the client
                # applies it and re-edits; nothing is silently clobbered
                return {"type": "reject", "field": field, "rev": current_rev,
                        "value": (row or {}).get("values", {}).get(field)}
            saved = await engine.draft_store.save_fields(
                s, rdef=rdef, defn=defn, resource_id=resource_id,
                part_key=part_key, audience=audience, fields={field: value},
                base_version=instance.version, actor=principal,
                at=engine.invoker.clock())
    return {"type": "saved", "field": field,
            "rev": saved["revs"][field], "saved_at": saved["saved_at"].isoformat(),
            "_value": value}


async def serve(engine: Any, ws: Any, rdef: Any, defn: Any,
                resource_id: str, part_key: str, principal: Any) -> None:
    """Drive one participant's connection for the lifetime of the socket."""
    from starlette.websockets import WebSocketDisconnect

    from .render import probe_transition

    rooms: CollabRooms = engine.collab
    key: RoomKey = (rdef.kind, resource_id, defn.name, part_key)
    await ws.accept()
    # affordance gate (design §11): the channel belongs to principals the
    # action currently renders for — visibility alone is not enough
    async with engine.storage.session() as s:
        instance = await engine.storage.load(s, rdef.kind, resource_id)
        if instance is None:
            await ws.close(code=4404)
            return
        ctx = engine.invoker._ctx(principal, s, mode="probe")
        if instance.state not in defn.from_:
            await ws.close(code=4403)
            return
        status, _, _, _ = await probe_transition(defn, instance, ctx)
        if status != "available":
            await ws.close(code=4403)
            return
        row = await engine.draft_store.load(
            s, rdef, defn, resource_id, part_key, audience_of(defn, principal))
    room = rooms.room(key)
    room.members[ws] = principal
    await ws.send_text(json.dumps({
        "type": "state",
        "values": (row or {}).get("values", {}),
        "revs": {k: int(v) for k, v in (row or {}).get("revs", {}).items()},
        "saved_at": row["saved_at"].isoformat() if row else None,
        "stale": (row["base_version"] != instance.version) if row else False,
        "participants": rooms.participants(key),
    }, default=str))
    await rooms.broadcast(key, {
        "type": "presence", "event": "joined", "actor": actor_of(principal),
        "participants": rooms.participants(key)}, exclude=ws)
    known = set(rdef.action_schemas[defn.name][0].get("properties", {}))
    try:
        while True:
            raw = await ws.receive_text()
            try:
                msg = json.loads(raw)
            except json.JSONDecodeError:
                await ws.send_text(json.dumps({
                    "type": "error", "errors": {"_root": ["invalid JSON"]}}))
                continue
            if not isinstance(msg, dict) or msg.get("type") != "update" \
                    or not isinstance(msg.get("field"), str):
                await ws.send_text(json.dumps({"type": "error", "errors": {
                    "_root": ["expected {\"type\": \"update\", \"field\", "
                              "\"value\", \"base_rev\"}"]}}))
                continue
            field = msg["field"]
            if field not in known:
                # same contract as a draft PUT: fields must be a subset of
                # the action's schema; values may be invalid mid-edit
                await ws.send_text(json.dumps({"type": "error", "errors": {
                    field: ["unknown field"]}}))
                continue
            base_rev = msg.get("base_rev")
            if not isinstance(base_rev, int):
                base_rev = 0
            outcome = await _drain(engine, rdef, defn, resource_id, part_key,
                                   field, msg.get("value"), base_rev,
                                   principal, room)
            if outcome is None:
                await rooms.close(key, "gone")
                return
            if outcome["type"] == "reject":
                await ws.send_text(json.dumps(outcome, default=str))
                continue
            value = outcome.pop("_value")
            await ws.send_text(json.dumps(outcome, default=str))
            relay = {"type": "update", "field": field, "value": value,
                     "rev": outcome["rev"], "actor": actor_of(principal),
                     "saved_at": outcome["saved_at"]}
            await rooms.broadcast(key, relay, exclude=ws)
            # cross-worker liveness rides the bus; our own echo is skipped
            await engine.draft_store.announce(
                rdef=rdef, defn=defn, resource_id=resource_id,
                part_key=part_key, frame=relay)
    except WebSocketDisconnect:
        pass
    finally:
        room.members.pop(ws, None)
        if room.members:
            await rooms.broadcast(key, {
                "type": "presence", "event": "left",
                "actor": actor_of(principal),
                "participants": rooms.participants(key)})
        else:
            rooms.rooms.pop(key, None)

File: waymark10/src/waymark10/server/collab.clj
(ns waymark10.server.collab
  "Live collab (phase 9b): waymark-relay over http-kit websockets at
  GET /api/{plural}/{id}/-/{action}/draft/collab, for :edit actions
  whose :draft is {:shared true :live true}. Clients join a per-draft
  room; frames:

    client → server   {\"type\": \"set\", \"field\", \"value\"}
    server → others   {\"type\": \"update\", \"field\", \"value\",
                       \"rev\", \"author\"}
    client → server   {\"type\": \"sync\"}
    server → sender   {\"type\": \"sync\", \"values\", \"rev\"}
    server → sender   {\"type\": \"error\", \"errors\": {field: [msgs]}}

  Every accepted set is the SAME write a draft PUT lands — partial
  validation against the action's input schema (unknown fields and
  type errors answer an error frame, values may be incomplete
  mid-edit), then the shared draft row persists through the drafts
  storage in the set's own transaction. The room's rev is a per-draft
  atom, incremented per accepted set and stamped on the broadcast.

  Merge discipline, scoped honestly to per-field last-writer-wins:
  server-ordered (the room lock serializes sets), no OT/CRDT, and —
  unlike waymark9's relay/2 — no base_rev staleness rejection, no
  saved/reject acks (sync is the read-back), no presence frames, no
  affordance gate or regate (the resource and a live shared draft
  policy are the door; what the action may DO stays the invoke's
  question), no cross-worker bus (rooms are process-local), and no
  closed frame when the draft is consumed — the next set answers an
  error and the client re-syncs. Each is a recorded punt, not a
  forgotten one.

  Rooms clean up on last disconnect; the rev resets with the room —
  LWW ordering only matters among live participants, and a rejoined
  room starts fresh over the persisted values."
  (:require [org.httpkit.server :as http]
            [waymark10.schema :as schema]
            [waymark10.server.problems :as p]
            [waymark10.server.store :as store]
            [waymark10.wire :as wire]))

(set! *warn-on-reflection* true)

;; ── rooms ───────────────────────────────────────────────────────────

(defn- room-of [eng key]
  (get (swap! (:collab-rooms eng) update key
              #(or % {:clients (atom {})
                      :rev (atom 0)
                      :lock (Object.)}))
       key))

(defn- leave! [eng key room ch]
  (let [clients (swap! (:clients room) dissoc ch)]
    (when (empty? clients)
      (swap! (:collab-rooms eng)
             (fn [rooms]
               ;; only reap the room we joined — a racing rejoin may
               ;; have re-minted the key
               (if (identical? (get rooms key) room)
                 (dissoc rooms key)
                 rooms))))))

(defn- send-frame! [ch frame]
  (http/send! ch (wire/write-json frame)))

(defn- broadcast!
  "Every room member but the sender."
  [room frame sender]
  (doseq [ch (keys @(:clients room))
          :when (not (identical? ch sender))]
    (send-frame! ch frame)))

(defn- author-of [principal]
  {:id (:id principal)
   :type (name (:type principal :human))
   :display (or (:display principal) (:id principal))})

;; ── the door ────────────────────────────────────────────────────────

(defn- live-defn
  "The action's normalized definition when its draft policy is
  {:shared true :live true}; 404 otherwise — a collab route for an
  unlive draft does not exist."
  [rdef action-name]
  (or (when-some [a (get-in rdef [:actions action-name])]
        (when (and (get-in a [:edit :draft :shared])
                   (get-in a [:edit :draft :live]))
          (assoc a :name action-name)))
      (throw (p/problem :not-found 404 "Not found"
                        {:detail (str (name (:kind rdef))
                                      " has no live shared draft on "
                                      (name action-name) ".")}))))

;; ── frames ──────────────────────────────────────────────────────────

(defn- error! [ch errors]
  (send-frame! ch {:type "error" :errors (p/wire-value errors)}))

(defn- load-shared-draft [eng kind id action-name]
  (store/with-tx (:storage eng)
    (fn [tx] (store/load-draft (:storage eng) tx kind id
                               action-name "shared"))))

(defn- persist-set!
  "The set's write: merge the field into the shared draft's stored
  values and upsert, base_version = the row's current version — the
  same row a draft PUT writes. → the merged values, or nil when the
  resource is gone (consumed rows outlive their drafts, vanished rows
  do not)."
  [eng rdef action-name id field value]
  (store/with-tx (:storage eng)
    (fn [tx]
      (when-some [raw (store/load-row (:storage eng) tx (:kind rdef) id {})]
        (let [draft (store/load-draft (:storage eng) tx (:kind rdef) id
                                      action-name "shared")
              values (assoc (or (:values draft) {}) field value)]
          (store/save-draft! (:storage eng) tx (:kind rdef) id
                             action-name "shared" values (:version raw))
          values)))))

(defn- handle-set!
  [eng rdef defn' id room ch msg principal]
  (let [field-str (:field msg)
        field (when (string? field-str) (keyword field-str))
        known (set (schema/entry-keys (:input defn')))]
    (cond
      (nil? field)
      (error! ch {:_root ["expected {\"type\": \"set\", \"field\", \"value\"}"]})

      (not (contains? known field))
      (error! ch {field ["unknown field"]})

      :else
      (let [value (:value msg)
            decoded (schema/decode (:input defn') {field value})]
        (if-some [errors (schema/partial-closed-errors (:input defn') decoded)]
          (error! ch errors)
          ;; the room lock serializes sets — server-ordered LWW
          (locking (:lock room)
            (if (nil? (persist-set! eng rdef (:name defn') id field value))
              (error! ch {:_root [(str "no " (name (:kind rdef)) " " id)]})
              (let [rev (swap! (:rev room) inc)]
                (broadcast! room {:type "update"
                                  :field field-str
                                  :value value
                                  :rev rev
                                  :author (author-of principal)}
                            ch)))))))))

(defn- handle-sync!
  [eng rdef defn' id room ch]
  (let [draft (load-shared-draft eng (:kind rdef) id (:name defn'))]
    (send-frame! ch {:type "sync"
                     :values (p/wire-value (or (:values draft) {}))
                     :rev @(:rev room)})))

(defn- handle! [eng rdef defn' id room ch raw principal]
  (let [msg (when (string? raw)
              (try (wire/read-json raw) (catch Exception _ ::bad)))]
    (cond
      (or (= ::bad msg) (not (map? msg)))
      (error! ch {:_root ["invalid JSON"]})

      (= "set" (:type msg))
      (handle-set! eng rdef defn' id room ch msg principal)

      (= "sync" (:type msg))
      (handle-sync! eng rdef defn' id room ch)

      :else
      (error! ch {:_root ["expected type \"set\" or \"sync\""]}))))

;; ── the route ───────────────────────────────────────────────────────

(defn join
  "Upgrade one request into a room membership for the draft's
  lifetime. The resource must exist and the action must declare a
  live shared draft — refusals throw problems BEFORE the upgrade, so
  the handshake answers them as plain HTTP."
  [eng rdef action-name id principal req]
  (let [defn' (live-defn rdef action-name)
        _ (or (store/with-tx (:storage eng)
                (fn [tx] (store/load-row (:storage eng) tx (:kind rdef) id {})))
              (throw (p/not-found (:kind rdef) id)))
        key [(:kind rdef) id (:name defn')]
        room (room-of eng key)]
    (http/as-channel
     req
     {:on-open (fn [ch]
                 (swap! (:clients room) assoc ch (author-of principal)))
      :on-receive (fn [ch raw]
                    (handle! eng rdef defn' id room ch raw principal))
      :on-close (fn [ch _status]
                  (leave! eng key room ch))})))
